tsl/src/chunk.c

```diff
@@ -186,3 +206,889 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type, boo
 
 	return num_results;
 }
+
+#if PG16_LT
+/* VacuumCutoffs is defined in PG16 and greater. Define here for previous PG
+ * versions. */
+struct VacuumCutoffs
+{
+	TransactionId relfrozenxid;
+	MultiXactId relminmxid;
+	TransactionId OldestXmin;
+	MultiXactId OldestMxact;
+	TransactionId FreezeLimit;
+	MultiXactId MultiXactCutoff;
+};
+#endif
+
+typedef struct RelationMergeInfo
+{
+	Oid relid;
+	struct VacuumCutoffs cutoffs;
+	const Chunk *chunk;
+	Relation rel;
+} RelationMergeInfo;
+
+typedef enum MergeLockUpgrade
+{
+	MERGE_LOCK_UPGRADE,
+	MERGE_LOCK_CONDITIONAL_UPGRADE,
+	MERGE_LOCK_ACCESS_EXCLUSIVE,
+} MergeLockUpgrade;
+
+static void
+compute_rel_vacuum_cutoffs(Relation rel, struct VacuumCutoffs *cutoffs)
```

Comment:
Why not move it to compat.h and make it vacuum_get_cutoffs_compat to follow the same pattern of other backward compatibility code?

Reply:
Moved some of this to compat.h, but decided to refactor a bit to implement the new API in old versions instead of the other way around.